_posts/2017-11-16-totw-99.md

```diff
@@ -0,0 +1,239 @@
+definition. When evaluating an API, we must consider definitions beyond the
+class body that can be as much a part of its interface as its public members.
+
+These external interface points include template specializations like hashers or
+traits, nonmember operator overloads (e.g. logging, relationals), and other
+canonical nonmember functions designed for use with argument-dependent lookup
+(ADL), most notably `swap`.
+
+A few of these are illustrated below for some sample class `space::Key`:
+
+```c++
+namespace space {
+class Key { ... };
+bool operator==(const Key& a, const Key& b);
+bool operator<(const Key& a, const Key& b);
+void swap(Key& a, Key& b);
+// standard streaming
+std::ostream& operator<<(std::ostream& os, const Key& x);
+// gTest printing
+void PrintTo(const Key& x, std::ostream* os);
+// new-style flag extension:
+bool ParseFlag(const string& text, Key* dst, string* err);
+string UnparseFlag(const Key& v);
+}  // namespace space
+
+HASH_NAMESPACE_BEGIN
+template <>
+struct hash<space::Key> {
+  size_t operator()(const space::Key& x) const;
+};
+HASH_NAMESPACE_END
+```
+
+There are some important risks associated with making such extensions
+incorrectly, so this article will try to present some guidance.
+
+### The proper namespace
+
+Interface points that are functions are usually designed to be found by
+argument-dependent lookup (ADL, see [TotW 49](/tips/49)). Operators and some
+operator-like functions (notably `swap`) are designed to be found by ADL. This
+protocol only works reliably when the function is defined in a namespace
+associated with the type being customized. The associated namespaces include
+those of its base classes and class template parameters. A common mistake is to
+place these functions in the global namespace. To illustrate the problem,
+consider the following code in which `good(x)` and `bad(x)` functions are called
+with identical syntax:
+
+```c++
+namespace library {
+struct Letter {};
+void good(Letter);
+}  // namespace library
+
+// bad is improperly placed in global namespace
+void bad(library::Letter);
+
+namespace client {
+void good();
+void bad();
+
+void test(const library::Letter& x) {
+  good(x);  // ok: 'library::good' is found by ADL.
+  bad(x);  // oops: '::bad' is hidden by 'client::bad'.
+}
+}  // namespace client
+```
+
+Note the difference between `library::good` and `::bad`. The `test` function is
+relying on the absence of any function called `bad` in namespaces enclosing the
+call site. The appearance of `client::bad` hides `::bad` from the `test` caller.
+Meanwhile, the `good` function is found regardless of what else exists in the
+`test` function's enclosing scope. The C++ name lookup sequence will only yield
+a global if a name search from closer lexical scopes to the call site fails to
+find any names.
+
+This is all very subtle, which is the point, really. It's all much simpler if we
+default to defining functions alongside the data on which they operate.
```

Comment:
Worth emphasizing that last line?

*default to defining functions alongside the data on which they operate*